server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -0,0 +1,371 @@
+            } else {
+                numIndicesUnchanged++;
+            }
+            previousStateIndexMetadataVersionByName.remove(indexMetadata.getIndex().getName());
+        }
+
+        for (String removedIndexName : previousStateIndexMetadataVersionByName.keySet()) {
+            allUploadedIndexMetadata.remove(removedIndexName);
+        }
+        final ClusterMetadataMarker marker = uploadMarker(
+            clusterState,
+            allUploadedIndexMetadata.values().stream().collect(Collectors.toList()),
+            false
+        );
+        final long durationMillis = relativeTimeMillisSupplier.getAsLong() - startTimeMillis;
+        if (durationMillis >= slowWriteLoggingThreshold.getMillis()) {
+            logger.warn(
+                "writing cluster state took [{}ms] which is above the warn threshold of [{}]; "
+                    + "wrote  metadata for [{}] indices and skipped [{}] unchanged indices",
+                durationMillis,
+                slowWriteLoggingThreshold,
+                numIndicesUpdated,
+                numIndicesUnchanged
+            );
+        } else {
+            // todo change to debug
+            logger.info(
+                "writing cluster state took [{}ms]; " + "wrote and metadata for [{}] indices and skipped [{}] unchanged indices",
+                durationMillis,
+                numIndicesUpdated,
+                numIndicesUnchanged
+            );
+        }
+        return marker;
+    }
+
+    @Nullable
+    public ClusterMetadataMarker markLastStateAsCommitted(ClusterState clusterState, ClusterMetadataMarker previousMarker)
+        throws IOException {
+        if (clusterState.nodes().isLocalNodeElectedClusterManager() == false) {
+            logger.error("Local node is not elected cluster manager. Exiting");
+            return null;
+        }
+        assert clusterState != null : "Last accepted cluster state is not set";
+        assert previousMarker != null : "Last cluster metadata marker is not set";
+        assert REMOTE_CLUSTER_STATE_ENABLED_SETTING.get(settings) == true : "Remote cluster state is not enabled";
+        return uploadMarker(clusterState, previousMarker.getIndices(), true);
+    }
+
+    public ClusterState getLatestClusterState(String clusterUUID) {
+        // todo
+        return null;
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (blobStoreRepository != null) {
+            IOUtils.close(blobStoreRepository);
+        }
+    }
+
+    // Visible for testing
+    void ensureRepositorySet() {
+        if (blobStoreRepository != null) {
+            return;
+        }
+        final String remoteStoreRepo = REMOTE_CLUSTER_STATE_REPOSITORY_SETTING.get(settings);
+        assert remoteStoreRepo != null : "Remote Cluster State repository is not configured";
+        final Repository repository = repositoriesService.get().repository(remoteStoreRepo);
```

Comment:
this will throw `RepositoryMissingException` and will cause cluster state publish to fail. We should handle this exception and currently we are only skipping remote cluster state upload if `STATE_NOT_RECOVERED_BLOCK` exists. But repo might not be available even in the next cluster state publish as next cluster state publish is registering the repo?

Just want to make sure by assuming that after STATE_NOT_RECOVERED_BLOCK is removed we will have repo registered might not lead to failures. Maybe we can introduce another block like `STATE_NOT_RECOVERED_BLOCK` for remote state repo registration in a subsequent PR?

Reply:
Let me handle the RepositoryMissingException for now. Once #9105 is pushed, we can revisit to modify the check for remote state upload.